File: typescript/punto4.ts
// Estructuras de Datos en TypeScript

// Arrays

let array: number[] = [3, 1, 4, 1, 5];
array.push(9); // Inserción
array.splice(1, 1); // Borrado
array[2] = 2; // Actualización
array.sort((a, b) => a - b); // Ordenación


// Sets

let set: Set<number> = new Set([3, 1, 4, 1, 5]);
set.add(9); // Inserción
set.delete(1); // Borrado


// Maps

let map: Map<string, number> = new Map([
    ["uno", 1],
    ["dos", 2]
]);
map.set("tres", 3); // Inserción
map.delete("dos"); // Borrado
map.set("tres", 33); // Actualización


// Stack (Pila) con Array

class Stack<T> {
    private items: T[] = [];
    push(item: T) { this.items.push(item); }
    pop(): T | undefined { return this.items.pop(); }
}


// Queue (Cola) con Array

class Queue<T> {
    private items: T[] = [];
    enqueue(item: T) { this.items.push(item); }
    dequeue(): T | undefined { return this.items.shift(); }
}


// Agenda de contactos en terminal

import * as readline from "readline";

interface Contact {
    name: string;
    phone: string;
}

class ContactBook {
    private contacts: Contact[] = [];

    addContact(name: string, phone: string) {
        if (!/^[0-9]{1,11}$/.test(phone)) {
            console.log("Número de teléfono inválido.");
            return;
        }
        this.contacts.push({ name, phone });
        console.log("Contacto agregado.");
    }

    updateContact(name: string, newPhone: string) {
        let contact = this.contacts.find(c => c.name === name);
        if (contact) contact.phone = newPhone;
        else console.log("Contacto no encontrado.");
    }

    deleteContact(name: string) {
        this.contacts = this.contacts.filter(c => c.name !== name);
        console.log("Contacto eliminado.");
    }

    searchContact(name: string) {
        let contact = this.contacts.find(c => c.name === name);
        if (contact) console.log(contact);
        else console.log("No encontrado.");
    }

    listContacts() {
        console.log("Contactos:", this.contacts);
    }
}

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
const contactBook = new ContactBook();

function mainMenu() {
    rl.question("Elige operación (add, update, delete, search, list, exit): ", (option) => {
        switch (option) {
            case "add":
                rl.question("Nombre: ", name => {
                    rl.question("Teléfono: ", phone => {
                        contactBook.addContact(name, phone);
                        mainMenu();
                    });
                });
                break;
            case "update":
                rl.question("Nombre: ", name => {
                    rl.question("Nuevo Teléfono: ", phone => {
                        contactBook.updateContact(name, phone);
                        mainMenu();
                    });
                });
                break;
            case "delete":
                rl.question("Nombre: ", name => {
                    contactBook.deleteContact(name);
                    mainMenu();
                });
                break;
            case "search":
                rl.question("Nombre: ", name => {
                    contactBook.searchContact(name);
                    mainMenu();
                });
                break;
            case "list":
                contactBook.listContacts();
                mainMenu();
                break;
            case "exit":
                rl.close();
                break;
            default:
                console.log("Opción no válida");
                mainMenu();
        }
    });
}

mainMenu();

File: golang/punto4.go
package main

import (
	"fmt"
	"sort"
)

// Listas (Slices)
func listOperations() {
	array := []int{3, 1, 4, 1, 5}
	array = append(array, 9)                // Inserción
	array = append(array[:1], array[2:]...) // Borrado
	array[1] = 2                            // Actualización
	sort.Ints(array)                        // Ordenación
	fmt.Println("Lista ordenada:", array)
}

// Conjuntos (Maps como Sets)
func setOperations() {
	setData := map[int]bool{3: true, 1: true, 4: true, 5: true}
	setData[9] = true  // Inserción
	delete(setData, 1) // Borrado
	fmt.Println("Conjunto:", setData)
}

// Diccionarios (Maps)
func mapOperations() {
	mapData := map[string]int{"uno": 1, "dos": 2}
	mapData["tres"] = 3    // Inserción
	delete(mapData, "dos") // Borrado
	mapData["tres"] = 33   // Actualización
	fmt.Println("Mapa:", mapData)
}

// Pila (Stack) con Slice
type Stack struct {
	items []int
}

func (s *Stack) Push(item int) {
	s.items = append(s.items, item)
}

func (s *Stack) Pop() int {
	if len(s.items) == 0 {
		return -1
	}
	item := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return item
}

//s := Stack{}
//s.Push(10)
//s.Push(20)
//fmt.Println(s.Pop()) // 20
//fmt.Println(s.Pop()) // 10
//fmt.Println(s.Pop()) // -1 (stack is empty)

// Cola (Queue) con Slice
type Queue struct {
	items []int
}

func (q *Queue) Enqueue(item int) {
	q.items = append(q.items, item)
}

func (q *Queue) Dequeue() int {
	if len(q.items) == 0 {
		return -1
	}
	item := q.items[0]
	q.items = q.items[1:]
	return item
}

//q := Queue{}
//q.Enqueue(10)
//q.Enqueue(20)
//fmt.Println(q.Dequeue()) // 10
//fmt.Println(q.Dequeue()) // 20
//fmt.Println(q.Dequeue()) // -1 (queue is empty)

func main() {

}
